### ISSUE TITLE
Docs build on Windows not working

### ISSUE THREAD

**Issue by tcmetzger** [bokeh/bokeh#13544]
Building the docs on Windows doesn't seem to work. I have cloned a fresh copy of the repo on Win10 (about 14 days ago) and set everything up as described in our [contributor guide](https://docs.bokeh.org/en/latest/docs/dev_guide/documentation.html#setting-up-and-building-bokeh-s-documentation). When I run `make.bat html`, the following things happen (tested on two separate Win10 machines and with both branch-3.4 and branch-3.3 - with the same outcome):

```
(bkdev) C:\github\bokeh\docs\bokeh>make.bat html
Running Sphinx v7.2.6
C:\github\bokeh\src\bokeh\sphinxext\bokeh_gallery.py:44: RemovedInSphinx80Warning: The alias 'sphinx.util.status_iterator' is deprecated, use 'sphinx.util.display.status_iterator' instead. Check CHANGES for Sphinx API modifications.
  from sphinx.util import ensuredir, status_iterator
C:\github\bokeh\src\bokeh\sphinxext\bokeh_plot.py:104: RemovedInSphinx80Warning: The alias 'sphinx.util.status_iterator' is deprecated, use 'sphinx.util.display.status_iterator' instead. Check CHANGES for Sphinx API modifications.
  from sphinx.util import copyfile, ensuredir, status_iterator
C:\github\bokeh\src\bokeh\sphinxext\bokeh_sitemap.py:41: RemovedInSphinx80Warning: The alias 'sphinx.util.status_iterator' is deprecated, use 'sphinx.util.display.status_iterator' instead. Check CHANGES for Sphinx API modifications.
  from sphinx.util import status_iterator
making output directory... done
creating gallery file entries... [100%] missing_dates.rst
[autosummary] generating autosummary for: docs/dev_guide.rst, docs/dev_guide/bindings.rst, docs/dev_guide/bokehjs.rst, docs/dev_guide/creating_issues.rst, docs/dev_guide/debug_headless_chrome.rst, docs/dev_guide/documentation.rst, docs/dev_guide/env_vars.rst, docs/dev_guide/models.rst, docs/dev_guide/pull_requests.rst, docs/dev_guide/python.rst, ..., docs/user_guide/topics/contour.rst, docs/user_guide/topics/geo.rst, docs/user_guide/topics/graph.rst, docs/user_guide/topics/hex.rst, docs/user_guide/topics/hierarchical.rst, docs/user_guide/topics/images.rst, docs/user_guide/topics/pie.rst, docs/user_guide/topics/stats.rst, docs/user_guide/topics/timeseries.rst, index.rst
BokehDeprecationWarning: 'tile_providers module' was deprecated in Bokeh 3.0.0 and will be removed, use 'add_tile directly' instead.
loading intersphinx inventory from https://numpy.org/doc/stable/objects.inv...
loading intersphinx inventory from https://pandas.pydata.org/pandas-docs/stable/objects.inv...
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://www.sphinx-doc.org/en/master/objects.inv...
loading intersphinx inventory from https://xyzservices.readthedocs.io/en/stable/objects.inv...
building [mo]: targets for 0 po files that are out of date
writing output...
building [html]: targets for 610 source files that are out of date
updating environment: [new config] 610 added, 0 changed, 0 removed
 * Serving Flask app 'bokeh_app_9b25ccc505b0478d9f5ba38c8c3ac53a'
 * Debug mode: off
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5050
Press CTRL+C to quit
```
-> The docs begin to build but at some point the server started (I only called `make.bat html`, NOT  `make.bat serve`). When I open the URL in my browser, I only see a 404 error.

Next, I hit Ctrl+C to quit the server. The following output happens:

```
127.0.0.1 - - [08/Nov/2023 16:29:02] "GET / HTTP/1.1" 404 -
 * Serving Flask app 'bokeh_app_fd60335a86254100a2a6631df1f58797'ents_source
 * Debug mode: off
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5050
Press CTRL+C to quit
```
i.e. the server just restarted (still 404). I hit CTRL+C again, and the following happens:

```
copying bokeh-plot files... [100%] bokeh-content-fbe67463385d4aabb3ee0fe6f09fd2c2-docs-examples-basic-bars-bar.js
adding links to sitemap...

Sphinx error:
Error generating bokeh-content-2f8ebb4ba27240b6be1c650f99705f86-docs-examples-basic-layouts-custom_layout.js:

bokeh-plot:: error:

Traceback (most recent call last):
  File "C:\github\bokeh\src\bokeh\application\handlers\code_runner.py", line 229, in run
    exec(self._code, module.__dict__)
  File "C:\github\bokeh\examples\basic\layouts\custom_layout.py", line 61, in <module>
    save([p0, p1, p2], template=template)
  File "C:\github\bokeh\src\bokeh\sphinxext\example_handler.py", line 89, in _add_root
    curdoc().add_root(obj)
  File "C:\github\bokeh\src\bokeh\document\document.py", line 321, in add_root
    with self.models.freeze():
  File "C:\Users\Timo Metzger\anaconda3\envs\bkdev\lib\contextlib.py", line 142, in __exit__
    next(self.gen)
  File "C:\github\bokeh\src\bokeh\document\models.py", line 135, in freeze
    self._pop_freeze()
  File "C:\github\bokeh\src\bokeh\document\models.py", line 288, in _pop_freeze
    self.recompute()
  File "C:\github\bokeh\src\bokeh\document\models.py", line 215, in recompute
    new_models |= mr.references()
AttributeError: 'list' object has no attribute 'references'


evaluating source:

from bokeh.io import save
from bokeh.plotting import figure
from bokeh.util.browser import view

template = """
{% block preamble %}
<style>
* { box-sizing: border-box; }
.plots { display: flex; flex-direction: row; width: 100%; }
.p { width: 33.3%; padding: 50px; }
.p:nth-child(1) { background-color: red; }
.p:nth-child(2) { background-color: green; }
.p:nth-child(3) { background-color: blue; }
</style>
{% endblock %}
{% block body %}
<body style="background-color: lightgray;">
    {{ self.inner_body() }}
</body>
{% endblock %}
{% block contents %}
<div>
<p>This example shows how different Bokeh Document roots may be embedded in custom
templates. The individual plots were embedded in divs using the embed macro:

<pre>
    &lt;div class="p"&gt;&#123;&#123; embed(roots.p0) &#125;&#125;&lt;/div&gt;
    &lt;div class="p"&gt;&#123;&#123; embed(roots.p1) &#125;&#125;&lt;/div&gt;
    &lt;div class="p"&gt;&#123;&#123; embed(roots.p2) &#125;&#125;&lt;/div&gt;
</pre>

And the divs are styled using standard CSS in the template:

<pre>
    .p { width: 33.3%; padding: 50px; }
    .p:nth-child(1) { background-color: red; }
    .p:nth-child(2) { background-color: green; }
    .p:nth-child(3) { background-color: blue; }
</pre>
</p>
</div>
<div class="plots">
    <div class="p">{{ embed(roots.p0) }}</div>
    <div class="p">{{ embed(roots.p1) }}</div>
    <div class="p">{{ embed(roots.p2) }}</div>
</div>
</div>
{% endblock %}
"""

x = [1, 2, 3]
y = [1, 2, 3]

p0 = figure(name="p0", sizing_mode="scale_width")
p0.scatter(x, y, size=20, fill_color="red")
p1 = figure(name="p1", sizing_mode="scale_width")
p1.scatter(x, y, size=20, fill_color="green")
p2 = figure(name="p2", sizing_mode="scale_width")
p2.scatter(x, y, size=20, fill_color="blue")

save([p0, p1, p2], template=template)
view("custom_layout.html")

File not found - static
0 File(s) copied
Terminate batch job (Y/N)? n
```

It seems like several strange things are happening here... 

I usually build the docs in WSL - I would expect `make.bat html` to work just like  `make html`  works in WSL!

cc @bryevdv @vicadesoba

**Comment by huijeong1015** [comment]
I was investigating this issue to work on #11765, and these are the problems that I've found. I'm new to this project, so I don't think I have the expertise to fix this issue, but hope these information helps.

**1. Flask run issue**
In the `examples/basic/data/ajax_source.py` and `examples/basic/data/server_sent_events_source.py`, they have `app.run(port=5050)` lines called without `if __name__ == '__main__':`. These two files result in those two Flask servers that need to be quit with CTRL+C.

**2. AttributeError: 'list' object has no attribute 'reference'**
This issue is due to line 61 of `examples/basic/layouts/custom_layout.py`: `save([p0, p1, p2], template=template)`. 
This line calls `_add_root` function in `src/bokeh/sphinxext/example_handler.py` with a list of models as `obj`. This results in adding a list of models into the `document.roots`, which throws the above error at `new_models |= mr.references()` line in `src/bokeh/document/models.py` because `mr` points to the list instead of `Model` class.

**3. `view` function in `examples/basic/layouts/custom_layout.py`**
I think this function should also be enclosed with `if __name__ == '__main__':`. Since we run the `make.bat html` from the `docs/bokeh` path, it creates a problem when opening the `custom_layout.html` file, which gets created under `examples/basic/layouts/`.

**Comment by bryevdv** [comment]
Thanks for the information @huijeong1015. @vicadesoba has reported seeing the same issues. From these descriptions, it is curious why things would work with the real Makefile on linux or OSX, but they clearly do as those systems are utilized to build docs continuously. @vicadesoba has also reported that building the docs does work on windows under WSL (which uses the real Makefile). 

To be honest, at this point I would just advocate for deleting `make.bat` and stating that Windows docs builds are only supported via WSL or linux VM, using the real Makefile. I would like for a "native Windows" option to exist, ideally, but I am just not sure it is worth sinking more effort into, in practice.

**Comment by tcmetzger** [comment]
To get this discussion going again, here is a completely different idea: 

I have created a branch called `timo/docker-experiments` that adds a customized [dev container](https://code.visualstudio.com/docs/devcontainers/create-dev-container) for docs building and testing. Basically, by adding a [.devconatainer.json file](https://github.com/bokeh/bokeh/blob/timo/docker-experiments/.devcontainer/devcontainer.json) to the repo, VS code automatically builds a Docker container with a Linux-based Bokeh dev environment and all env variables set up for building and serving docs (see https://containers.dev/overview for context).

To test this, you need to make sure you have Docker running on your system (e.g. [Docker desktop](https://www.docker.com/products/docker-desktop)) and you have the [Dev Containers extension](Dev Containers extension) installed in VSCode. Git checkout the branch `timo/docker-experiments` and start vscode on this branch (`code .`). Once VS Code loads, it should ask you if you want to reload from the dev env. If you click yes, VS Code will automatically set up the container, download and install all dependencies, and set up the env variables (this takes several minutes). Once that is done, you can open a new Terminal, which will automatically be a shell inside the Linux container. There, you can run things like `make html serve` (the server's port from the container will automatically be forwarded so that you can use any browser on your host system).

![image](https://github.com/bokeh/bokeh/assets/39711796/4daaf7eb-e518-41b2-8327-872c4cde6660)


This works surprisingly well, and I have been using this to work on docs on my Windows machine. This setup provides a very consistent, replicable environment. My branch includes a shell script that automatically picks the most recent conda environment YML file, so there shouldn't be too much maintenance burden.

An obvious drawback is that this only works with VS Code (and IntelliJ, according to https://containers.dev/supporting). Also, Docker is required, which might still be a hurdle for new contributors. There also can be some difficulties related to using GH credentials from the host system inside the Docker container. This whole setup definitely needs some more testing, and we might come to the conclusion that this is not worth pursuing at all - in the end, making the make.bat work again might be a more efficient solution!

**Comment by mosc9575** [comment]
I was reading the post above very carefully and I recognized that all files which are causing problems are on the `bokeh_sampledata_xref_skiplist` in the conf.py.

https://github.com/bokeh/bokeh/blob/78d71c26208ca81fb151904cff25d7cbf01f5f1b/docs/bokeh/source/conf.py#L123-L130

I am not sure if I am write but I guess there is just a simple problem comparing paths. If I remeber correctly these files were causing problems on linux, too and as a outcome the `bokeh_sampledata_xref_skiplist` was added.

I go for it an try to solve the issue.

Thanks to all of you for the affort you put in to this issue, which looks like it is based on my earlier implementation.

Edit:

As a reference, the list was added in [#12144](https://github.com/bokeh/bokeh/pull/12114) and there are some comments addressing the issue.
